src/absil/illib.fs

```diff
@@ -503,9 +503,19 @@ module String =
     let uppercase (s: string) =
         s.ToUpperInvariant()
 
-    let isUpper (s: string) = 
-        s.Length >= 1 && Char.IsUpper s.[0] && not (Char.IsLower s.[0])
-        
+    // Scripts that distinguish between upper and lower case (bicameral) DU Discriminators and Active Pattern identifiers are required to start with an upper case character.
+    // For valid identifiers where the case of the identifier can not be determined because there is no upper and lower case we will allow DU Discriminators and upper case characters 
+    // to be used.  This means that developers using unicameral scripts such as hindi, are not required to prefix these identifiers with an Upper case latin character. 
+    //
+    let isLeadingIdentifierCharacterUpperCase (s:string) =
+        let isUpperCaseCharacter c =
+            // if IsUpper and IsLower return the same value, then we can't tell if it's upper or lower case, so it is a case insensensitive language
+            // thusly the char must be good.
+            if Char.IsUpper c = Char.IsLower c then Char.IsLetter c
+            else Char.IsUpper c
```

Comment:
Sorry for suggesting another option, just thought it could be better to call `IsUpper` once and, well, first, since it's what we're looking for in the first place. 🙂 
```suggestion
            Char.IsUpper c || Char.IsLetter c && not Char.IsLower c
```

Reply:
Is it defined that `IsUpper` implies `IsLetter` always?

Reply:
```csharp
```
https://github.com/dotnet/runtime/blob/87f78fe5424ace6e3de3693d0fda5cd25011cb1d/src/libraries/System.Private.CoreLib/src/System/Char.cs#L272-L278

Reply:
@cloudRoutine 
For sure this is easier to describe and reason about:
````
            if Char.IsUpper c = Char.IsLower c then Char.IsLetter c
            else Char.IsUpper c
````
than
````
Char.IsUpper c || Char.IsLetter c && not Char.IsLower c
```` 
Also inlining and hoisting IsUpper, is something I believe the jit should do, although I don't know if it does.  And finally, if for reasons, I can't even fathom an implementation of a script caused IsUpper and IsLower to both return true then we are back where we are now.  But TBH the readability of the current impl wins for me.

Reply:
> And finally, if for reasons, I can't even fathom an implementation of a script caused IsUpper and IsLower to both return true then we are back where we are now. 

@KevinRansom The Unicode standard has characters where both properties are true. However, .NET doesn't follow those rules strictly and returns `false` and `false` for these cases:

![image](https://user-images.githubusercontent.com/16015770/82123217-7612fb00-9798-11ea-8526-5a512d5053ab.png)

```f#
> Char.IsUpper(char(0x02B0));;
val it : bool = false

> Char.IsLower(char(0x02B0));;
val it : bool = false

> Array.init 65535 id |> Array.tryFind(fun x -> Char.IsUpper(char x) && Char.IsLower(char x));;
val it : int option = None
```

Reply:
> For sure this is easier to describe and reason about:

You can prevent an extra call to `GetUnicodeCategory` (which is relatively expensive) by changing it to the following, which is still readable:

```f#
let isUpper = Char.IsUpper c
if isUpper = Char.IsLower c then Char.IsLetter c
else isUpper
```

Or, this is arguable more readable and also has max 3 checks for the UnicodeCategory.

```f#
match Char.IsUpper c, Char.IsLower c with
| true, true
| false, false -> Char.IsLetter c
| isUpper, _ -> isUpper
````

Reply:
I will make it this:
````
let isUpper = Char.IsUpper c
if isUpper = Char.IsLower c then Char.IsLetter c
else isUpper
````